### PR TITLE
[Graph C API] Registration API

### PR DESCRIPTION
**Pull request by Solaryee**
This PR is to add graph registration C API, which includes:
1. Registration C API in `tensorflow/c/experimental/grappler/`.
2. `PluginGraphOptimizerRegistry` in grappler to register and create optimizer.

This PR is following RFC [Modular TensorFlow Graph C API](https://github.com/tensorflow/community/blob/master/rfcs/20201027-modular-tensorflow-graph-c-api.md).

### CODE REVIEW

**Comment by ShengYang1** [comment]
Hi @penpornk @ezhulenev,  please help to have a review. Thanks very much!

**Review by ezhulenev — APPROVED** [review]

**Comment by ShengYang1** [comment]
@ezhulenev MacOS UT failure is fixed, please review again. Thanks!

**Review by gbaned — APPROVED** [review]

**Comment by ShengYang1** [comment]
@ezhulenev @gbaned Seems the two MacOS UT failures are not related to this PR, one is about gRPC server, and the other is about python ImportError. Could you please help check? What should I do to fix the failures?

**Comment by penpornk** [comment]
@ShengYang1 It could be that the test instance was flaky. Let me rerun them.

File: tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc [hunk 1]
```diff
@@ -57,5 +97,113 @@ void CustomGraphOptimizerRegistry::RegisterOptimizerOrDie(
...
+  }
+  return optimizer_list;
+}
+
+void PluginGraphOptimizerRegistry::RegisterPluginOptimizerOrDie(
+    const Creator& optimizer_creator, const std::string& device_type,
+    ConfigsList& configs) {
+  auto ret = GetPluginConfigMap()->insert({device_type, configs});
+  if (!ret.second) {
+    LOG(FATAL) << "PluginGraphOptimizer with device_type " << device_type
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:120]: We don't allow `LOG(FATAL)` in new code. Can we just make TF warn that it finds a duplicate (which TF will ignore) and carry on without crashing? (Let's also remove `OrDie` from the function name.) If something can go wrong with this route, please explain why.

**Review by penpornk — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc (hunk 1 above)
Comment (penpornk) [reply tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:120]: Taking back my comment. We agree it's okay to fail. Please ignore this. :)

**Review by penpornk — APPROVED** [review]

File: tensorflow/c/experimental/grappler/grappler.cc [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      return tensorflow::Status(tensorflow::error::FAILED_PRECONDITION, \
+                                "'" #NAME "' field in " #STRUCT_NAME    \
+                                " must be set.");                       \
+    }                                                                   \
+  } while (0)
+
+tensorflow::Status ValidateTPOptimizerRegistrationParams(
+    const TP_OptimizerRegistrationParams& params) {
+  VALIDATE_STRUCT_SIZE(TP_OptimizerRegistrationParams, params,
+                       TP_OPTIMIZER_REGISTRARION_PARAMS_STRUCT_SIZE);
```
Comment (penpornk) [inline tensorflow/c/experimental/grappler/grappler.cc:59]: Please fix in other places as well.
```suggestion
                       TP_OPTIMIZER_REGISTRATION_PARAMS_STRUCT_SIZE);
```

**Review by penpornk — CHANGES_REQUESTED** [review]

File: tensorflow/c/experimental/grappler/grappler.cc (hunk 2 above)
Comment (ShengYang1) [reply tensorflow/c/experimental/grappler/grappler.cc]: Done.

**Review by penpornk — APPROVED** [review]
Thank you for the quick fixes!

**Comment by penpornk** [comment]
Just a heads up that I've made several changes internally before the merge. Most of them won't have any effects on other code. There are two changes that can break other codes using this PR:
- Renamed `ConfigsList` to [ConfigList].(https://github.com/tensorflow/tensorflow/blob/88affffa8fc78966b4623f1e6e8c98045f94abfd/tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.h#L30)
- Changed the signature of `BufferToMessage` from
```c++
Status BufferToMessage(const TF_Buffer* in,
                       tensorflow::protobuf::MessageLite& out);
```
to
```c++
Status BufferToMessage(const TF_Buffer* in,
                       tensorflow::protobuf::MessageLite* out);
```

Thank you again for the PR!

**Comment by penpornk** [comment]
One more minor change: Renamed the versioning variables from `SE_MAJOR`, etc. to `GO_MAJOR`, etc. https://github.com/tensorflow/tensorflow/commit/bca7074097cd86333cfb809e8b15d881afd83b69

**Comment by EmployedRussian** [comment]
FYI, this code:

```
PluginRegistrationMap* registered_plugin_optimizers = nullptr;
PluginRegistrationMap* GetPluginRegistrationMap() {
  if (registered_plugin_optimizers == nullptr)
    registered_plugin_optimizers = new PluginRegistrationMap;
  return registered_plugin_optimizers;
}
```
is thread-unsafe (two threads calling `GetPluginRegistrationMap()` in parallel will race on `registered_plugin_optimizers`) and may result in a memory leak.

Fix forthcoming.

**Pull request merged**